Mathlib/RingTheory/Trace/Quotient.lean

```diff
@@ -0,0 +1,311 @@
+  · intro x
+    obtain ⟨x, rfl⟩ := Ideal.Quotient.mk_surjective x
+    obtain ⟨x, s, rfl⟩ := IsLocalization.mk'_surjective p.primeCompl x
+    obtain ⟨s', hs⟩ := Ideal.Quotient.mk_surjective (I := p) (Ideal.Quotient.mk p s)⁻¹
+    simp only [IsScalarTower.algebraMap_eq R Rₚ (Rₚ ⧸ _),
+      Ideal.Quotient.algebraMap_eq, RingHom.comp_apply]
+    use x * s'
+    rw [← sub_eq_zero, ← map_sub, Ideal.Quotient.eq_zero_iff_mem]
+    have : algebraMap R Rₚ s ∉ maximalIdeal Rₚ := by
+      rw [← Ideal.mem_comap, IsLocalization.AtPrime.comap_maximalIdeal Rₚ p]
+      exact s.prop
+    refine ((inferInstanceAs <| (maximalIdeal Rₚ).IsPrime).mem_or_mem ?_).resolve_left this
+    rw [mul_sub, IsLocalization.mul_mk'_eq_mk'_of_mul, IsLocalization.mk'_mul_cancel_left,
+      ← map_mul, ← map_sub, ← Ideal.mem_comap, IsLocalization.AtPrime.comap_maximalIdeal Rₚ p,
+      mul_left_comm,
+      ← Ideal.Quotient.eq_zero_iff_mem, map_sub, map_mul, map_mul, hs, mul_inv_cancel, mul_one,
+      sub_self]
+    rw [Ne, Ideal.Quotient.eq_zero_iff_mem]
+    exact s.prop
+
+lemma IsLocalization.AtPrime.map_eq_maximalIdeal :
+    p.map (algebraMap R Rₚ) = maximalIdeal Rₚ := by
+  convert congr_arg (Ideal.map (algebraMap R Rₚ))
+    (IsLocalization.AtPrime.comap_maximalIdeal Rₚ p).symm
+  rw [map_comap p.primeCompl]
+
+local notation "pS" => Ideal.map (algebraMap R S) p
+local notation "pSₚ" => Ideal.map (algebraMap Rₚ Sₚ) (maximalIdeal Rₚ)
+
+lemma comap_map_eq_map_of_isLocalization_algebraMapSubmonoid :
+    (Ideal.map (algebraMap R Sₚ) p).comap (algebraMap S Sₚ) = pS := by
+  rw [IsScalarTower.algebraMap_eq R S Sₚ, ← Ideal.map_map, eq_comm]
+  apply Ideal.le_comap_map.antisymm
+  intro x hx
+  obtain ⟨α, hα, hαx⟩ : ∃ α ∉ p, α • x ∈ pS := by
+    have ⟨⟨y, s⟩, hy⟩ := (IsLocalization.mem_map_algebraMap_iff
+      (Algebra.algebraMapSubmonoid S p.primeCompl) Sₚ).mp hx
+    rw [← map_mul,
+      IsLocalization.eq_iff_exists (Algebra.algebraMapSubmonoid S p.primeCompl)] at hy
+    obtain ⟨c, hc⟩ := hy
+    obtain ⟨α, hα, e⟩ := (c * s).prop
+    refine ⟨α, hα, ?_⟩
+    rw [Algebra.smul_def, e, Submonoid.coe_mul, mul_assoc, mul_comm _ x, hc]
+    exact Ideal.mul_mem_left _ _ y.prop
+  obtain ⟨β, γ, hγ, hβ⟩ : ∃ β γ, γ ∈ p ∧ β * α = 1 + γ := by
+    obtain ⟨β, hβ⟩ := Ideal.Quotient.mk_surjective (I := p) (Ideal.Quotient.mk p α)⁻¹
+    refine ⟨β, β * α - 1, ?_, ?_⟩
+    · rw [← Ideal.Quotient.eq_zero_iff_mem, map_sub, map_one,
+        map_mul, hβ, inv_mul_cancel, sub_self]
+      rwa [Ne, Ideal.Quotient.eq_zero_iff_mem]
+    · rw [add_sub_cancel]
+  have := Ideal.mul_mem_left _ (algebraMap _ _ β) hαx
+  rw [← Algebra.smul_def, smul_smul, hβ, add_smul, one_smul] at this
+  refine (Submodule.add_mem_iff_left _ ?_).mp this
+  rw [Algebra.smul_def]
+  apply Ideal.mul_mem_right
+  exact Ideal.mem_map_of_mem _ hγ
+
+variable (S Sₚ)
+
+/-- The isomorphism `S ⧸ pS ≃+* Sₚ ⧸ pSₚ`. -/
+noncomputable
+def quotMapEquivQuotMapMaximalIdealOfIsLocalization : S ⧸ pS ≃+* Sₚ ⧸ pSₚ := by
```

Comment:
I think this def and the ones (in this section) before it should also go somewhere else.